File: pkg/@eser/cache/cache.test.ts
// Copyright 2023-present Eser Ozvataf and other contributors. All rights reserved. Apache-2.0 license.

import * as assert from "@std/assert";
import * as standardsRuntime from "@eser/standards/runtime";
import { createCacheManager } from "./cache.ts";

const createTempCacheManager = async () => {
  const tempDir = await standardsRuntime.runtime.fs.makeTempDir({
    prefix: "cache-test-",
  });

  return {
    cache: createCacheManager({
      app: { name: "test-app", org: "test-org" },
      baseDir: tempDir,
    }),
    tempDir,
    cleanup: async () => {
      try {
        await standardsRuntime.runtime.fs.remove(tempDir, { recursive: true });
      } catch {
        // Ignore cleanup errors
      }
    },
  };
};

Deno.test("createCacheManager returns CacheManager", () => {
  const cache = createCacheManager({
    app: { name: "test-app" },
  });

  assert.assertExists(cache.getCacheDir);
  assert.assertExists(cache.getVersionedPath);
  assert.assertExists(cache.exists);
  assert.assertExists(cache.ensureDir);
  assert.assertExists(cache.list);
  assert.assertExists(cache.remove);
  assert.assertExists(cache.clear);
});

Deno.test("getCacheDir returns valid path", () => {
  const cache = createCacheManager({
    app: { name: "test-app" },
  });

  const cacheDir = cache.getCacheDir();

  assert.assertExists(cacheDir);
  assert.assert(cacheDir.length > 0, "Cache dir should not be empty");
  assert.assert(
    cacheDir.includes("test-app"),
    "Cache dir should include app name",
  );
});

Deno.test("getCacheDir with org includes org in path", () => {
  const cache = createCacheManager({
    app: { name: "test-app", org: "test-org" },
  });

  const cacheDir = cache.getCacheDir();

  assert.assert(
    cacheDir.includes("test-org"),
    "Cache dir should include org name",
  );
  assert.assert(
    cacheDir.includes("test-app"),
    "Cache dir should include app name",
  );
});

Deno.test("getVersionedPath normalizes version", () => {
  const cache = createCacheManager({
    app: { name: "test-app" },
  });

  const path = cache.getVersionedPath("1.0.0", "binary");

  assert.assert(
    path.includes("v1.0.0"),
    "Should normalize version to have v prefix",
  );
  assert.assert(path.includes("binary"), "Should include item name");
});

Deno.test("ensureDir creates directory", async () => {
  const { cache, cleanup } = await createTempCacheManager();

  try {
    const testDir = standardsRuntime.runtime.path.join(
      cache.getCacheDir(),
      "subdir",
    );

    await cache.ensureDir(testDir);

    const dirExists = await standardsRuntime.runtime.fs.exists(testDir);
    assert.assert(dirExists, "Directory should exist after ensureDir");
  } finally {
    await cleanup();
  }
});

Deno.test("exists returns false for non-existent path", async () => {
  const { cache, cleanup } = await createTempCacheManager();

  try {
    const result = await cache.exists("non-existent-file");

    assert.assertEquals(result, false);
  } finally {
    await cleanup();
  }
});

Deno.test("exists returns true for existing path", async () => {
  const { cache, cleanup } = await createTempCacheManager();

  try {
    // Create a test file
    const testDir = cache.getCacheDir();
    await standardsRuntime.runtime.fs.mkdir(testDir, { recursive: true });

    const testFile = standardsRuntime.runtime.path.join(testDir, "test-file");
    await standardsRuntime.runtime.fs.writeTextFile(testFile, "test content");

    const result = await cache.exists("test-file");

    assert.assertEquals(result, true);
  } finally {
    await cleanup();
  }
});

Deno.test("list returns empty array for empty cache", async () => {
  const { cache, cleanup } = await createTempCacheManager();

  try {
    const entries = await cache.list();

    assert.assertExists(entries);
    assert.assertEquals(entries.length, 0);
  } finally {
    await cleanup();
  }
});

Deno.test("list returns entries for non-empty cache", async () => {
  const { cache, cleanup } = await createTempCacheManager();

  try {
    const cacheDir = cache.getCacheDir();
    await standardsRuntime.runtime.fs.mkdir(cacheDir, { recursive: true });

    // Create test files
    await standardsRuntime.runtime.fs.writeTextFile(
      standardsRuntime.runtime.path.join(cacheDir, "file1"),
      "content1",
    );
    await standardsRuntime.runtime.fs.mkdir(
      standardsRuntime.runtime.path.join(cacheDir, "dir1"),
    );

    const entries = await cache.list();

    assert.assertEquals(entries.length, 2);

    const fileEntry = entries.find((e) => e.name === "file1");
    const dirEntry = entries.find((e) => e.name === "dir1");

    assert.assertExists(fileEntry);
    assert.assertExists(dirEntry);
    assert.assertEquals(fileEntry?.isDirectory, false);
    assert.assertEquals(dirEntry?.isDirectory, true);
  } finally {
    await cleanup();
  }
});

Deno.test("remove deletes file", async () => {
  const { cache, cleanup } = await createTempCacheManager();

  try {
    const cacheDir = cache.getCacheDir();
    await standardsRuntime.runtime.fs.mkdir(cacheDir, { recursive: true });

    const testFile = standardsRuntime.runtime.path.join(cacheDir, "to-remove");
    await standardsRuntime.runtime.fs.writeTextFile(testFile, "content");

    // Verify file exists
    assert.assertEquals(await cache.exists("to-remove"), true);

    // Remove it
    await cache.remove("to-remove");

    // Verify it's gone
    assert.assertEquals(await cache.exists("to-remove"), false);
  } finally {
    await cleanup();
  }
});

Deno.test("clear removes all cache contents", async () => {
  const { cache, cleanup } = await createTempCacheManager();

  try {
    const cacheDir = cache.getCacheDir();
    await standardsRuntime.runtime.fs.mkdir(cacheDir, { recursive: true });

    // Create some files
    await standardsRuntime.runtime.fs.writeTextFile(
      standardsRuntime.runtime.path.join(cacheDir, "file1"),
      "content",
    );
    await standardsRuntime.runtime.fs.writeTextFile(
      standardsRuntime.runtime.path.join(cacheDir, "file2"),
      "content",
    );

    // Clear cache
    await cache.clear();

    // Verify cache dir is gone
    const dirExists = await standardsRuntime.runtime.fs.exists(cacheDir);
    assert.assertEquals(dirExists, false);
  } finally {
    await cleanup();
  }
});

File: pkg/ajan/cachefx/cache_test.go
// Copyright 2023-present Eser Ozvataf and other contributors. All rights reserved. Apache-2.0 license.

package cachefx_test

import (
	"os"
	"path/filepath"
	"strings"
	"testing"

	"github.com/eser/stack/pkg/ajan/cachefx"
)

func TestNewManagerDir(t *testing.T) {
	t.Parallel()

	t.Run("uses BaseDir when provided", func(t *testing.T) {
		t.Parallel()

		m := cachefx.NewManager(cachefx.Options{
			BaseDir: "/tmp/testcache",
			App:     cachefx.AppIdentifier{Name: "myapp"},
		})

		if m.Dir() != "/tmp/testcache/myapp" {
			t.Errorf("Dir() = %q, want %q", m.Dir(), "/tmp/testcache/myapp")
		}
	})

	t.Run("includes Org prefix when set", func(t *testing.T) {
		t.Parallel()

		m := cachefx.NewManager(cachefx.Options{
			BaseDir: "/tmp/testcache",
			App:     cachefx.AppIdentifier{Org: "eser", Name: "ajan"},
		})

		want := "/tmp/testcache/eser/ajan"
		if m.Dir() != want {
			t.Errorf("Dir() = %q, want %q", m.Dir(), want)
		}
	})

	t.Run("falls back to XDG cache home when BaseDir empty", func(t *testing.T) {
		t.Parallel()

		m := cachefx.NewManager(cachefx.Options{
			App: cachefx.AppIdentifier{Name: "myapp"},
		})

		// Should end with /myapp and be non-empty.
		if !strings.HasSuffix(m.Dir(), string(filepath.Separator)+"myapp") {
			t.Errorf("Dir() = %q, expected suffix /myapp", m.Dir())
		}
	})
}

func TestVersionedPath(t *testing.T) {
	t.Parallel()

	m := cachefx.NewManager(cachefx.Options{
		BaseDir: "/cache",
		App:     cachefx.AppIdentifier{Name: "app"},
	})

	cases := []struct {
		version  string
		name     string
		expected string
	}{
		{"1.2.3", "pkg.tar.gz", "/cache/app/v1.2.3/pkg.tar.gz"},
		{"v1.2.3", "pkg.tar.gz", "/cache/app/v1.2.3/pkg.tar.gz"}, // v prefix normalised
		{"0.0.1", "data", "/cache/app/v0.0.1/data"},
	}

	for _, tc := range cases {
		tc := tc
		t.Run(tc.version+"/"+tc.name, func(t *testing.T) {
			t.Parallel()

			got := m.VersionedPath(tc.version, tc.name)
			if got != tc.expected {
				t.Errorf("VersionedPath(%q, %q) = %q, want %q", tc.version, tc.name, got, tc.expected)
			}
		})
	}
}

func TestListEmptyDir(t *testing.T) {
	t.Parallel()

	m := cachefx.NewManager(cachefx.Options{
		BaseDir: "/nonexistent-base-dir-xyz",
		App:     cachefx.AppIdentifier{Name: "app"},
	})

	entries, err := m.List()
	if err != nil {
		t.Fatalf("unexpected error listing non-existent dir: %v", err)
	}

	if len(entries) != 0 {
		t.Errorf("expected empty slice, got %d entries", len(entries))
	}
}

func TestExistsEnsureDirRemoveClear(t *testing.T) {
	t.Parallel()

	tmp := t.TempDir()
	m := cachefx.NewManager(cachefx.Options{
		BaseDir: tmp,
		App:     cachefx.AppIdentifier{Name: "testapp"},
	})

	cacheDir := m.Dir()

	t.Run("Exists returns false for missing dir", func(t *testing.T) {
		if m.Exists(cacheDir) {
			t.Error("expected Exists=false before EnsureDir")
		}
	})

	t.Run("EnsureDir creates directory", func(t *testing.T) {
		if err := m.EnsureDir(cacheDir); err != nil {
			t.Fatalf("EnsureDir: %v", err)
		}

		if !m.Exists(cacheDir) {
			t.Error("expected Exists=true after EnsureDir")
		}
	})

	t.Run("List returns created entry after file write", func(t *testing.T) {
		if err := os.WriteFile(filepath.Join(cacheDir, "foo.txt"), []byte("hi"), 0o644); err != nil {
			t.Fatalf("write: %v", err)
		}

		entries, err := m.List()
		if err != nil {
			t.Fatalf("List: %v", err)
		}

		if len(entries) != 1 || entries[0].Name != "foo.txt" {
			t.Errorf("expected 1 entry 'foo.txt', got %v", entries)
		}
	})

	t.Run("Remove deletes specific file", func(t *testing.T) {
		if err := m.Remove(filepath.Join(cacheDir, "foo.txt")); err != nil {
			t.Fatalf("Remove: %v", err)
		}

		entries, _ := m.List()
		if len(entries) != 0 {
			t.Errorf("expected empty after Remove, got %d entries", len(entries))
		}
	})

	t.Run("Clear removes entire cache dir", func(t *testing.T) {
		_ = os.WriteFile(filepath.Join(cacheDir, "bar.txt"), []byte("bye"), 0o644)

		if err := m.Clear(); err != nil {
			t.Fatalf("Clear: %v", err)
		}

		if m.Exists(cacheDir) {
			t.Error("cache dir should not exist after Clear")
		}
	})
}

func TestXdgCacheHome(t *testing.T) {
	// t.Setenv is incompatible with t.Parallel — runs serially.
	// Env override must take precedence.
	t.Setenv("XDG_CACHE_HOME", "/custom/cache")

	got := cachefx.XdgCacheHome()
	if got != "/custom/cache" {
		t.Errorf("XdgCacheHome() = %q, want %q", got, "/custom/cache")
	}
}
